base/loading.jl

```diff
@@ -280,6 +293,46 @@ function include_dependency(path::AbstractString)
     return nothing
 end
 
+# used to track optional dependencies during precompilation
+const _optional_dependencies = String[]
+
+"""
+    register_optional_dependency(mod::String)
+
+In a module, declare that the module with the name `mod` is a dependency
+for precompilation; that is, the module will need to be recompiled if this
+optional module becomes available in the future.
+
+This is only needed if your module depends on a module that is not used via
+`require`, `using`, `import`, or `importall`. It has no effect outside
```

Comment:
Since `require` isn't exported, it might be better to prefix it with Base here for clarity.